modules/tensor_mechanics/include/interfaces/GradientOperator.h

```diff
@@ -0,0 +1,30 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "Moose.h"
+#include "MooseTypes.h"
+#include "libmesh/vector_value.h"
+#include "RankTwoTensor.h"
+
+class GradientOperator
+{
+protected:
+  /// Returns the gradient operator in a given direction
+  virtual RankTwoTensor
+  gradOp(unsigned int component, const RealVectorValue & g, const Real & v, const Point & p);
+
+  /// Accumulate the contribution of the gradient operator in a given direction for a given tensor
+  virtual void addGradOp(RankTwoTensor & G,
```

Comment:
Those don't need to be virtual as far as I can see.

Reply:
simply removing the virtual won't be any good of course, but this could be replaced with compile time resolution using templates.

Reply:
```
#include <iostream>

template <class T>
class X : public T {
	public:
		void run() {
			this->output();
		}
};

class A {
	public:
		void output()
		{
			std::cout << "A\n";
		}
};


class B {
	public:
		void output()
		{
			std::cout << "B\n";
		}
};


int main()
{
	X<A> a;
	a.run();
	X<B> b;
	b.run();
	return 0;
}
```

instead of 

```
#include <iostream>

class X {
	public:
		void run() {
			this->output();
		}
		virtual void output() = 0;
};

class A : public X {
	public:
		void output() override
		{
			std::cout << "A\n";
		}
};


class B : public X {
	public:
		void output() override
		{
			std::cout << "B\n";
		}
};


int main()
{
	A a;
	a.run();
	B b;
	b.run();
	return 0;
}
```

Reply:
Ughh, it's a pain to name them.